src/NServiceBus.Core/Transports/Development/DevelopmentTransportMessagePump.cs

```diff
@@ -0,0 +1,306 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Extensibility;
+    using Logging;
+    using Transport;
+
+    class DevelopmentTransportMessagePump : IPushMessages
+    {
+        public DevelopmentTransportMessagePump(string basePath)
+        {
+            this.basePath = basePath;
+        }
+
+        public Task Init(Func<MessageContext, Task> onMessage, Func<ErrorContext, Task<ErrorHandleResult>> onError, CriticalError criticalError, PushSettings settings)
+        {
+            this.onMessage = onMessage;
+            this.onError = onError;
+            transactionMode = settings.RequiredTransactionMode;
+
+            path = Path.Combine(basePath, settings.InputQueue);
+            var delayedRootPath = Path.Combine(path, ".delayed");
+
+            Directory.CreateDirectory(Path.Combine(path, ".committed"));
+            Directory.CreateDirectory(delayedRootPath);
+
+            delayedRootDirectory = new DirectoryInfo(delayedRootPath);
+
+            purgeOnStartup = settings.PurgeOnStartup;
+
+            receiveCircuitBreaker = new RepeatedFailuresOverTimeCircuitBreaker("DevelopmentTransportReceive", TimeSpan.FromSeconds(30), ex => criticalError.Raise("Failed to receive from " + settings.InputQueue, ex));
+
+            delayedMessagePoller = new Timer(MoveDelayedMessagesToMainDirectory);
+            return TaskEx.CompletedTask;
+        }
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            runningReceiveTasks = new ConcurrentDictionary<Task, Task>();
+            concurrencyLimiter = new SemaphoreSlim(limitations.MaxConcurrency);
+            cancellationTokenSource = new CancellationTokenSource();
+
+            cancellationToken = cancellationTokenSource.Token;
+
+            if (purgeOnStartup)
+            {
+                Array.ForEach(Directory.GetFiles(path), File.Delete);
+            }
+
+            messagePumpTask = Task.Factory.StartNew(() => ProcessMessages(), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
+
+            delayedMessagePoller.Change(TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(-1));
+        }
+
+        public async Task Stop()
+        {
+            cancellationTokenSource.Cancel();
+
+            // ReSharper disable once MethodSupportsCancellation
+            var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
```

Comment:
Should we use the same approach rabbit uses

```
while (semaphore.CurrentCount != maxConcurrency)
{
    await Task.Delay(50).ConfigureAwait(false);
}
```

and not track the currently running tasks since it simplifies the code?